File: data-structures/doubly-linked-list/typescript/doubly-linked-list.ts
class Node<T> {
	constructor(
		public value: T | null = null,
		public next: Node<T> | null = null,
		public prev: Node<T> | null = null
	) {}
}

export class LinkedList<T> {
	head: Node<T> | null;
	tail: Node<T> | null;
	length: number;

	constructor() {
		this.head = null;
		this.tail = null;
		this.length = 0;
	}

	push(value: T): void {
		const newNode = new Node(value);
		if (this.head === null) {
			this.head = newNode;
			this.tail = newNode;
		} else {
			newNode.prev = this.tail;
			this.tail!.next = newNode;
			this.tail = newNode;
		}
		this.length++;
	}

	insert(index: number, value: T): void {
		if (index <= 0) {
			return this.unshift(value);
		} else if (index > this.length - 1) {
			return this.push(value);
		}

		const newNode = new Node(value);

		const previousNode = this.get(index - 1);
		const currentNode = previousNode!.next;

		previousNode!.next = newNode;
		currentNode!.prev = newNode;

		newNode!.next = currentNode;
		newNode!.prev = previousNode;

		this.length++;
	}

	unshift(value: T): void {
		const newNode = new Node(value);
		if (this.length === 0) {
			this.head = newNode;
			this.tail = newNode;
		} else {
			this.head!.prev = newNode;
			newNode.next = this.head;
			this.head = newNode;
		}

		this.length++;
	}

	shift(): T | null {
		if (this.length === 0) {
			return null;
		}
		const returnedNode = this.head;
		if (this.length === 1) {
			this.head = null;
			this.tail = null;
		} else {
			const nextNode = this.head!.next;
			this.head = nextNode;
			nextNode!.prev = null;
			returnedNode!.next = null;
		}

		this.length--;
		return returnedNode!.value;
	}

	pop(): T | null {
		if (!this.tail) {
			return null;
		}

		let removedNode = this.tail;

		if (this.length === 1) {
			this.head = null;
			this.tail = null;
		} else {
			const previousNode = this.tail.prev;
			this.tail = previousNode;
			this.tail!.next = null;
			removedNode.prev = null;
		}

		this.length--;
		return removedNode.value;
	}

	remove(index: number): T | null {
		if (this.length === 0) {
			return null;
		}

		if (index <= 0) {
			return this.shift();
		} else if (index >= this.length - 1) {
			return this.pop();
		}
		const previousNode = this.get(index - 1);
		const removedNode = previousNode!.next;
		const nextNode = removedNode!.next;

		previousNode!.next = nextNode;
		nextNode!.prev = previousNode;

		removedNode!.next = null;
		removedNode!.prev = null;

		this.length--;
		return removedNode!.value;
	}

	get(index: number): Node<T> | null {
		if (index < 0 || index > this.length - 1) {
			throw new Error('Index out of bounds.');
		} else if (index === 0) {
			return this.head;
		} else if (index === this.length - 1) {
			return this.tail;
		}

		let currentNode: Node<T>;
		if (index < this.length / 2) {
			currentNode = this.head!;
			for (let i = 0; i < index; i++) {
				currentNode = currentNode!.next!;
			}
		} else {
			currentNode = this.tail!;
			for (let i = this.length - 1; i > index; i--) {
				currentNode = currentNode!.next!;
			}
		}

		return currentNode!;
	}

	setValue(index: number, value: T): Node<T> {
		const currentNode = this.get(index);
		currentNode!.value = value;
		return currentNode!;
	}

	printList(): T[] {
		const arrayList: T[] = [];
		let currentNode = this.head;
		while (currentNode !== null) {
			arrayList.push(currentNode.value as T);
			currentNode = currentNode.next;
		}
		return arrayList;
	}

	reverse(): void {
		if (this.length <= 1) {
			return;
		}

		let currentNode = this.head;
		let previousNode = null;
		let nextNode = null;

		[this.head, this.tail] = [this.tail, this.head];

		while (currentNode) {
			nextNode = currentNode.next;
			currentNode.next = previousNode;
			currentNode.prev = nextNode;
			previousNode = currentNode;
			currentNode = nextNode;
		}
	}
}

File: data-structures/doubly-linked-list/python/doubly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next: Node | None = None
        self.prev: Node | None = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, value):
        new_node = Node(value)
        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1

    def insert(self, index: int, value):
        if index <= 0:
            return self.prepend(value)
        elif index > self.length - 1:
            return self.append(value)

        new_node = Node(value)

        previous_node = self.get(index - 1)
        current_node = previous_node.next

        previous_node.next = new_node
        new_node.next = current_node
        new_node.prev = previous_node
        current_node.prev = new_node

        self.length += 1

    def prepend(self, value):
        new_node = Node(value)

        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node

        self.length += 1

    def pop_head(self):
        if self.length == 0:
            return None

        removed_node = self.head

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            next_node = self.head.next
            self.head = next_node

            next_node.prev = None
            removed_node.next = None

        self.length -= 1

        return removed_node.value

    def pop(self):
        if self.tail is None:
            return None

        removed_node = self.tail

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            new_tail = self.tail.prev
            new_tail.next = None
            self.tail = new_tail

        removed_node.prev = None
        self.length -= 1
        return removed_node.value

    def remove(self, index: int):
        if self.length == 0:
            return None

        if index <= 0:
            return self.pop_head()
        elif index >= self.length - 1:
            return self.pop()

        previous_node = self.get(index - 1)
        removed_node = previous_node.next
        next_node = removed_node.next

        previous_node.next = next_node
        next_node.prev = previous_node

        removed_node.prev = None
        removed_node.next = None

        self.length -= 1
        return removed_node.value

    def get(self, index: int) -> Node:
        if index < 0 or index > self.length - 1:
            raise IndexError("Index out of bounds.")

        if index == 0:
            return self.head
        elif index == self.length - 1:
            return self.tail

        if index < self.length / 2:
            current_node = self.head
            for _ in range(index):
                current_node = current_node.next
        else:
            current_node = self.tail
            for _ in range(self.length - 1, index, -1):
                current_node = current_node.prev

        return current_node

    def set_value(self, index: int, value) -> Node:
        current_node = self.get(index)
        current_node.value = value
        return current_node

    def print_list(self) -> list:
        printed_list = []
        current_node = self.head
        while current_node is not None:
            printed_list.append(current_node.value)
            current_node = current_node.next

        return printed_list

    def reverse(self):
        if self.length <= 1:
            return

        current_node = self.head
        previous_node = None

        self.head, self.tail = self.tail, self.head

        while current_node is not None:
            next_node = current_node.next
            current_node.next = previous_node
            current_node.prev = next_node
            previous_node = current_node
            current_node = next_node
